internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1068,122 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkOldSnapshotsPipelineRun sorts all snapshots for application and cancells all running interationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as cancelled run finally to be gracefuly cancelled
+// god please forgive me for this monstrosity of cognitive complexity 40+
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+	snapshots, err := a.loader.GetAllSnapshots(a.context, a.client, application)
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	// check for older snapshots, their current status and type
+	latestSnapshot := sortedSnapshots[0]
+	for i := 1; i < len(sortedSnapshots); i++ {
+
+		if gitops.IsGroupSnapshot(&sortedSnapshots[i]) {
+			// check for pr-group-sha
+			fmt.Println("Yes its group snapshot")
+			latestGroupSha, _ := gitops.GetPRGroupFromSnapshot(&latestSnapshot)
+			groupSha, _ := gitops.GetPRGroupFromSnapshot(&sortedSnapshots[i])
+			if groupSha == latestGroupSha {
+				// this is the snapshot to which we can cancel pipelinerun
+				// set this snapshot as cancelled, superseded
+				err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+				if err != nil {
+					a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", a.snapshot.Name)
+					return err
+				}
+				// now get all integration pipelineruns for a snapshot
+				integrationTestPipelineruns, err := getAllIntegrationPipelineRunsForSnapshot(a.context, a.client, &sortedSnapshots[i])
+				if err != nil {
+					a.logger.Error(err, "Failed to get all integration pipelineruns for snapshot", "snapshot.Name", a.snapshot.Name)
+					return err
+				}
+				for _, plr := range integrationTestPipelineruns {
+					plr := plr
+					pipelinerunStatus, _, err := h.GetRunningIntegrationPipelineRunStatus(a.context, a.client, &plr)
+					if err != nil {
+						a.logger.Error(err, "Failed to get integration test pipelinerun status", "snapshot.Name", plr.Name)
+						return err
+					}
+					if pipelinerunStatus == intgteststat.IntegrationTestStatusInProgress {
+						//remove finalizer and cancel pipelinerun
+						err = h.RemoveFinalizerFromPipelineRun(a.context, a.client, a.logger, &plr, h.IntegrationPipelineRunFinalizer)
```

Comment:
I think we should check here whether we are removing finalizers from GH or GL, wdyt?